src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -216,45 +216,54 @@ internal sealed partial class VisualStudioProject
             _documentFileChangeContext.FileChanged += DocumentFileChangeContext_FileChanged;
         }
 
-        private void ChangeProjectProperty<T>(ref T field, T newValue, Func<Solution, Solution> withNewValue, bool logThrowAwayTelemetry = false)
+        private void ChangeProjectProperty<T>(ref T field, T newValue, Func<Solution, Solution> updateSolution, bool logThrowAwayTelemetry = false)
         {
-            using (_gate.DisposableWait())
-            {
-                ChangeProjectProperty_NoLock(ref field, newValue, withNewValue, logThrowAwayTelemetry);
-            }
+            ChangeProjectProperty(
+                ref field,
+                newValue,
+                (solutionChanges, oldValue) => solutionChanges.UpdateSolutionForProjectAction(Id, updateSolution(solutionChanges.Solution)),
+                logThrowAwayTelemetry);
         }
 
-        private void ChangeProjectProperty_NoLock<T>(ref T field, T newValue, Func<Solution, Solution> withNewValue, bool logThrowAwayTelemetry = false)
+        private void ChangeProjectProperty<T>(ref T field, T newValue, Action<SolutionChangeAccumulator, T> updateSolution, bool logThrowAwayTelemetry = false)
```

Comment:
can we just make this async? and push crazy waiting upwards?

Reply:
Not easily: this is used in a bunch of property setters, and this is also going to be called in synchronous paths for the legacy project system.

Reply:
I will push the sync crazy though into a helper method so this function looks saner though.

Reply:
Cleaned up a bit.